ops/main.py

```diff
@@ -169,6 +170,23 @@ def main(charm_class):
     # TODO: For Windows, when symlinks are used, this is not a valid
     #       method of getting an event name (see LP: #1854505).
     juju_exec_path = Path(sys.argv[0])
+    has_dispatch = juju_exec_path.name == 'dispatch'
+    if has_dispatch:
+        # The executable was 'dispatch', which means the actual hook we want to
+        # run needs to be looked up in the JUJU_DISPATCH_PATH env var, where it
+        # should be a path relative to the charm directory (the directory that
+        # holds `dispatch`). If that path actually exists, we need to exec to
+        # it instead of continuing.
+        #
+        # We don't look at the existence of JUJU_DISPATCH_PATH because we don't
+        # unset it on exec'ing into the hook, and it could be a symlink back to
+        # ourselves.
+        dispatch_path = Path(os.environ['JUJU_DISPATCH_PATH'])
+        juju_exec_path = juju_exec_path.parent / dispatch_path
+        if juju_exec_path.exists():
+            argv = sys.argv[:]
+            argv[0] = str(juju_exec_path)
+            subprocess.run(argv, check=True)
```

Comment:
We probably don't want to replace 'juju_exec_path' here, so we can do:
```
if dispatch_path.is_symlink() and dispatch_path.resolve() == juju_exec_path.resolve():
```
right?

Reply:
This feels like if the hook fails, the output becomes a Python traceback to indicate that the overall event has failed. Is there a way we could make that cleaner?

We already have a test that if, eg, hook/install returns nonzero that we treat the overall event as failed, correct?